### PR TITLE
Allow giving jobs names via execute for IBMQ devices

### LINKED ISSUE #3294: execute should take job_name as a kwarg
### What is the expected enhancement?
In the `ibmq-provider` version `0.4+` it is possible to assign a job a name via `backend.run(qobj, job_name)`.  It should be possible to assign a name via `execute` as well.

### PR DESCRIPTION
**Pull request by kdk**
Re-opens #3622.

Closes #3294

### CODE REVIEW

File: qiskit/execute.py [hunk 1]
```diff
@@ -242,4 +245,8 @@ def execute(experiments, backend,
                     )
 
     # executing the circuits on the backend and returning the job
-    return backend.run(qobj, **run_config)
+    if backend.name().split('_')[0] == 'ibmq':
```
Comment (kdk) [inline qiskit/execute.py]: Will this always be true? e.g. ibmqx2

Comment (nonhermitian) [reply qiskit/execute.py]: A good point. Will fix

**Comment by 1ucian0** [comment]
Dont forget about release notes!

File: qiskit/execute.py [hunk 2]
```diff
@@ -242,4 +245,8 @@ def execute(experiments, backend,
                     )
 
     # executing the circuits on the backend and returning the job
-    return backend.run(qobj, **run_config)
+    if type(backend).__name__ == 'IBMQBackend':
```
Comment (1ucian0) [inline qiskit/execute.py]: I think `isinstance` would be better here.

**Comment by 1ucian0** [comment]
`run_config` was designed for this kind of situations.

It seems to me that this is already supported: `execute(..., run_config={'job_name': job_name}`

File: qiskit/execute.py (hunk 2 above)
Comment (nonhermitian) [reply qiskit/execute.py]: But then I need to import the instance to check this and Terra is supposed to be independent of the IBM Q provider.

**Comment by nonhermitian** [comment]
`run_config` does work.  However if I change the backend to a simulator then it fails.  In addition, the user has no idea that this is even an option because there is nothing to tell you what `run_config` accepts.  As already mentioned, the acceptable kwargs that one can pass in this option are backend specific.

**Comment by 1ucian0** [comment]
> In addition, the user has no idea that this is even an option because there is nothing to tell you what run_config accepts. As already mentioned, the acceptable kwargs that one can pass in this option are backend specific.

That's true for all the kwargs in `run_config`. Probably the documentation can be improved, mentioning that the dict will be passed to `backend.run(..., **run_config)`.

**Comment by nonhermitian** [comment]
@1ucian0 I am not sure what the block is for here. This works, prevents errors due to backend specificity of the run_config kwargs, and documents the feature for end users.

**Comment by 1ucian0** [comment]
The full point of `run_config` is handling this situation.

> run_config does work. However if I change the backend to a simulator then it fails. 

This is what I would expected. In the same way that you changed the parameter `backend` in `execute`, you also need to change `run_config`.  It is analogous to `map`, which might fail if you change `func`. `run_config` is forward to `backend.run` as parameters. I agree that more documentation can clarify this. Alternatively, we can recommend backends to ignore kwargs that are not in their `run`, implementing something like `def run(qobj, ..., **_):`. That is a different discussion.

Asking Terra to support `job_name` is asking terra to administrate a namespace that belongs to the backend. What if in the future another backend supports `job_name` but with a different semantics? What if the backend renames or drops `job_name`?

It is true that we handle other backend-specific-ish parameters (like `seed_simulator`). However, they are part of the qobj specification. That's way they are supported for every backend. I do think these kind of parameters should be also pull out the qboj and being handle with `run_config` (again, different discussion). We should move in that direction of making Terra more backend-agnostic instead of the direction that this PR suggests.

TL;DR - My main reason for blocking is because I think this PR undermines a fundamental principle of Terra: backend independency.

**Comment by 1ucian0** [comment]
Do you agree?

**Comment by ajavadia** [comment]
Why can't this introduce an explicit `job_name` kwarg, but then have it be dumped into the `run_config` and shipped to the backend? The backend can decide what to do with it. Just leave a note saying that this option works if backend supports it, and IBM backends currently do.
The explicit hardcoding and special casing on `'IBMQBackend'` string type name is a bit weird.

BTW this is like every other option. Half of those options are not generic for every backend, they either work only for simulator backends, or pulse backends, etc.. `execute` is supposed to be a high level wrapper function, whoever uses it presumably understands the backend they are sending jobs to. Documentation is important.

**Comment by nonhermitian** [comment]
Because simulators will fail if it is added. So to get ibmq backend names to work that way the simulators and other providers need to be able to handle it gracefully. Execute should not fail if an additional option is passed that is irrelevant for success of the job itself. The string comparison is because terra does not depend on the ibmq provider, so I cannot import the instance to check with isinstance. With no documentation, run_config is impossible to rely on as there is no set of rules on how to use it or how it is consumed by various backends.

**Comment by ajavadia** [comment]
Yes my assumption is the user passes the option to backends that support it. This should of course be documented, but for example we currently can do:

execute(circuit, backend, meas_level=1)

There is no code in execute to catch that backend is definitely a pulse enabled backend. If backend is a circuit simulator, this will fail and we have been fine with that. I don’t see how this is different.

**Comment by 1ucian0** [comment]
Things likes `result = execute(circuit, BasicAer.get_backend('qasm_simulator'), meas_level=1)` do not fail. That is because `meas_level` ends up in qobj. I think backends ignore fields in the qobj that are not used by them. However, `job_name` is not like `meas_level`. See [`backend.run` signature](https://github.com/Qiskit/qiskit-ibmq-provider/blob/02851865a4140994f680a4adf442bbdf1740dfc9/qiskit/providers/ibmq/ibmqbackend.py#L73-L78).

The same behavior for `job_name` in `run_config` can be reached by asking backend to ignore kwargs that are not supported by them.

**Comment by ajavadia** [comment]
Ok based on the `IBMQBackend.run()` signature that you posted above, i'm not sure `run_config` will work for this case then. Run config is supposed to go in the qobj (by the assembler), but this is outside qobj. How does that work?

@1ucian0 said:
> It seems to me that this is already supported: execute(..., run_config={'job_name': job_name}

Regardless I think if I do `execute(circ, backend=simulator, job_name=...)`, this should be an error and not just silently ignored (as is done in this PR). I have made an assumption which cannot hold with my chosen backend.

**Comment by nonhermitian** [comment]
Setting `meas_level=2` is silently ignored by backends that do not support pulse, as is the case with many other options.  So not sure what changes with this PR in that respect.  In order for that to work, each backend must raise an error on any kwargs in `run_config` it does not recognize.  This will be pulse dependent on the IBMQ devices.

**Comment by ajavadia** [comment]
@nonhermitian you are right the current behavior is that irrelevant options are ignored. Your PR is thus keeping with the current behavior. 

I am just saying that this is not a good behavior in my opinion. We can't just ignore an option that the user set, without raising a warning or error to inform the user. So I would be fine if you just allow this to fail here for Aer backends (omit the backend check). And then we follow up later to also catch other invalid qobj options on client-side, before sending qobj.

Reopening PR.

**Comment by 1ucian0** [comment]
Since we agree the unused options should trigger error/warning, I would like to focus on the how. 

I have the strong opinion that this responsibility should be on the provider/backend side via `run_config` for the reasons exposed [here](https://github.com/Qiskit/qiskit-terra/pull/3625#issuecomment-566841984).

**Comment by nonhermitian** [comment]
I am completely fine if the backends / providers take this over. Any option set that is not used / and does not affect the outcome should emit a warning not error.

This also means that the backends should dynamically generate their docstrings to indicate which kwargs they support in run config, e.g. the pulse options. 

The downside to this is that changing between backends is not as simple as before

**Comment by 1ucian0** [comment]
> I am completely fine if the backends / providers take this over. Any option set that is not used / and does not affect the outcome should emit a warning not error.

I agree, a warning is the way to go.

> This also means that the backends should dynamically generate their docstrings to indicate which kwargs they support in run config, e.g. the pulse options.

What do you mean by "dynamically generate their docstring"? The docstring is [there already](https://github.com/Qiskit/qiskit-ibmq-provider/blob/02851865a4140994f680a4adf442bbdf1740dfc9/qiskit/providers/ibmq/ibmqbackend.py#L79-L94).

> The downside to this is that changing between backends is not as simple as before

Not fully sure what you mean here. Maybe an example?

**Comment by nonhermitian** [comment]
A backend tells you nothing about which kwargs it takes:
```python
Docstring:      Backend class interfacing with an IBMQ backend.
Init docstring:
Initialize remote backend for IBM Quantum Experience.

Args:
    configuration: configuration of backend.
    provider: provider.
    credentials: credentials.
    api: api for communicating with the Quantum Experience.
```

Before, it was possible to simply change a backend in `execute` and it would work without warnings or errors, e.g. like noise_model can be passed to a real device.  Now warnings are emitted, which breaks the seamlessness that we touted before.

**Comment by 1ucian0** [comment]
> A backend tells you nothing about which kwargs it takes:

The `run_config` dict is passed to `backend.run`, not to the backend constructor. We can mention that in the  `execute` docstring. Something like "check `backend.run` to see what arguments can go in `run_config`".

> Before, it was possible to simply change a backend in execute and it would work without warnings or errors, e.g. like noise_model can be passed to a real device. Now warnings are emitted, which breaks the seamlessness that we touted before.

A warning will allow the execution and an equally simple way.

Either unused options are silently ignored (which is something that @ajavadia does not like, and I tend to agree) by the backend or the backend emits a warning (which is something that you don't like). It has to be one of them, I don't see any other way around.

**Comment by nonhermitian** [comment]
The first item will be addressed with the change to configurable backends. 

I am fine with warnings, just pointing that out.

**Comment by 1ucian0** [comment]
In summary, we are agreeing that:
 - The jobs name option should be handle in execute via `run_config`.
 - The backend should warning when elements from `run_config` are not being used.
 - The backend should extend the docstring to be explicit on what are the `run_config` elements that are being used.
 - Same should apply to backend-specific params, like pulse options.

@ajavadia and @kdk, are you okey with this?

If so, I will move this issue to the providers repo, since there is nothing here to do on the terra side. Or am I missing something?

**Comment by nonhermitian** [comment]
I would say one issue is how to handle the kwargs that are already in execute. Everything will be grouped into a single backend_config, and warnings will be emitted for each kwarg that a backend does not support. For example, any open device will give many warnings for the pulse args in execute.

**Comment by 1ucian0** [comment]
> Everything will be grouped into a single `backend_config`

It's an option, but I don't feel very strongly about it because the current backend arguments are well defined by the qobj specification (while `run_config` are totally in the backend namespace).

> any open device will give many warnings for the pulse args in execute.

Maybe it is possible to group them in a single warning?

**Comment by nonhermitian** [comment]
So I guess my statement is that soon the backend calls will be
‘’’
backend = Backend(backend_config)
backend.run([circuits])
‘’’

So the default behavior will be a warning when the default execute options are not valid. In the case of open devices this means that every call to execute will give a warning. Are we comfortable with that?

**Comment by 1ucian0** [comment]
What about something like this `backend = Backend(backend_config, param_warning=False)`?

**Comment by ajavadia** [comment]
With the v2 provider interface the need for `execute()` will be reduced (it should really exist for backwards compat only). 

This particular feature can just exist in `backend.run(circuit, job_name)`. This makes sense to me as job name and many other options are features of particular backends. 

So I'm going to close this and let's make sure it works as expected with the new ibmq_backend.run() method.
